common/interactive-object/client.mjs

```diff
@@ -0,0 +1,45 @@
+// Copyright 2021 HITCON Online Contributors
+// SPDX-License-Identifier: BSD-2-Clause
+
+/**
+ * TODO: jsdoc
+ */
+class InteractiveObjectClientBaseClass {
+  /**
+   * Create the client side of the extension.
+   * @constructor
+   * @param {ClientExtensionHelper} helper - An extension helper object for
+   * servicing various functionalities of the extension.
+   * @param {MapCoord} initialPosition - TODO
+   * @param {Object} displayConfig - TODO
+   * @param {Function} interactFunction - The function to be called when the user tries to
+   * interact with this object (on click, keydown, signal, etc.)
+   */
+  constructor(helper, initialPosition, displayConfig, interactFunction) {
+    this.helper = helper;
+    this.mapCoord = initialPosition;
+    this.displayConfig = displayConfig;
+
+    for (const conf of displayConfig) {
+      const {zIndex, layerName, renderFunction, renderArgs} = conf;
+      this.helper.mapRenderer.registerCustomizedLayerToDraw(zIndex, layerName, renderFunction, renderArgs);
+    }
+
+    // interact on click
+    this.helper.inputManager.registerCanvasOnClickMapCoord((clickedMapCoord) => {
+      // determine whether this object is clicked
```

Comment:
Optional: One extra callback per NPC might slow down the client when we've a lot of NPCs. Consider having a `InteractiveObjectManager` that registers one single callback and lookup the callback through a Map. Similar optimization might be available for `registerCustomizedLayerToDraw` as well.

Reply:
I don't get the point of this optimization, maybe we can discuss it in the meeting.